### ISSUE TITLE
Show memos for outgoing transcations

### ISSUE THREAD

**Issue by sha-265** [michaelWuensch/BitBanana#99]
I'm using BitBanana with CLN.

Memos are shown only for incoming transactions but not for outgoing. Memos are missing in the history view, but also in the transaction's details modal.

Please add the memos for outgoing transactions also, in history and in the modal.

Thank you

**Comment by michaelWuensch** [comment]
Thanks for the issue.
I found and fixed the bug. It will be working in the next version!

**Comment by sha-265** [comment]
Thank you

**Comment by michaelWuensch** [comment]
This is fixed with 0.8.8 which got just released

**Comment by sha-265** [comment]
Thank you. When this version will be available on the Play store?

**Comment by michaelWuensch** [comment]
@sha-265 It is already live, but I did a partial rollout to only 30% of users. I'll observe crash reports and if nothing bad occurs, I'll roll out to 100 % in 2 days.

**Comment by sha-265** [comment]
@michaelWuensch, is it just me or the memo should appear for theses both entries, that belong to the same rebalancing transaction?

![image](https://github.com/user-attachments/assets/af9cd8e7-2cd4-441a-9758-e49796ead914)

**Comment by michaelWuensch** [comment]
@sha-265 yeah, I guess it should show there as well. I am reopening the issue for furter investigation.

Did anything change for you in 0.8.8 in regard to memos for outgoing transactions? Are some shown now?

**Comment by sha-265** [comment]
> Did anything change for you in 0.8.8 in regard to memos for outgoing transactions? Are some shown now?

Yes, memos are shown for some of the outgoing transactions, for other outgoing transactions memos are missing, but I'm not sure if it is because there is no memo for these transactions or the memo exists but not shown. The rebalance transaction is a transaction that I was pretty sure that memo exists but not shown.

**Comment by sha-265** [comment]
Just to confirm. I found another standard transactions (not rebalance) that the memo is missing. I compared the same transactions in Zeus wallet, and the memo appears there.

![image](https://github.com/user-attachments/assets/599ad87e-b302-48f2-81ce-63d7f66493dc)
![image](https://github.com/user-attachments/assets/5cab84fd-bd02-4507-a78c-20a89aec2ead)

**Comment by michaelWuensch** [comment]
Is there any pattern/difference you recognize?
Like is it a keysend transaction or maybe a transaction to a specific node type or version etc?

**Comment by sha-265** [comment]
It's not keysend transactions, but I can't find a pattern. One is the rebalance transaction, second is regular one, and the third is MPP transaction (that appears as three different transactions, instead of one, but this is another bug I guess).

**Comment by michaelWuensch** [comment]
Thanks for the info, I am wrapping up the next commit for this issue as we speak.
This should fix MPP transactions being shown as different payments. It also adds the description (and Payer note) for outgoing payments that paid a bolt 12 offer. These were not shown so far.
Furthermore I replaced some code that might have failed with a more robust one.
I hope all of these measures will finally fix the issue.
I'll let you know when the next version is out so you can test.

**Comment by michaelWuensch** [comment]
I just released v0.8.9, it should be available on Google Play soon.
Please test it and let me know if it worked ! :smiley:

**Comment by sha-265** [comment]
Hmm, it tried the new version, and I have major performance issue, and the transactions history list not loaded. And if it been loaded once, it loaded only incoming transactions for some reason 🫢

**Comment by michaelWuensch** [comment]
Damn, what version of core lightning are you running? I tested with 24.11. And do you have a huge transaction history?
My own node is LND, so I don't have a core lightning node with lots of transactions, just a small regtest node and there everything worked fine.

**Comment by michaelWuensch** [comment]
Ok, I'll see if I can test it tomorrow with 23.11 and find out whats going on.

**Comment by michaelWuensch** [comment]
Hmm, I am a bit lost right now.
I tried a core lightning 23.11.2 regtest node with 0.8.9 and made some transactions. Everything works. 
I have an updaid invoice, 3 incoming transactions from different nodes, an outgoing to another core lightning node, an outgoing to an LND node, a keysend transaction and an mpp transaction to a LND node.
Transaction history loads without any problems and the mpp transaction is displayed as a single transaction now.
On the one hand this is good news, but on the other hand it is bad, because it might also affect nodes running new versions of Core Lightning.

I see a few possibilities on how to go on from here:
1. You download the 0.8.8 version from github and install that, then you should be able to use it again.
2. I have to find a way to simulate hundreds and thousands of transactions without doing it manually. So I can see if there is a performance problem.
3. If you are willing to install non official releases, I can provide you some APKs where I revert parts of the code. This way we can test which code changes break the transaction history for your node.

**Comment by sha-265** [comment]
1. I did, and it's working fine again.
2. I don't know if it has it functionality but maybe try [Polar](https://github.com/jamaljsr/polar).
3. I can test it if you unable to reproduce this issue.

**Comment by michaelWuensch** [comment]
Thanks, yes, I already use Polar a lot.
Ok, I did some more tests today.
The description is not exposed in the API that is why I have to parse the bolt11 string for the description which needs some time.
I had a custom self written function to do this that I replaced with proper bolt11 parsing from a library in the last update.
I did this as maybe my own function fails in some cases and therefore the app might not show the description on some payments.
It turns out the new parsing is about 8 times slower than the custom one I made.
On my phone parsing for 1000 payments took about 1 second with the new one which isn't that bad, but maybe you use an old slow smartphone?
Anyway, I now changed the code so that it does not automatically parse all payments on load time, but only those that are actually displayed, which means it will do the parsing on demand while you scroll in the transaction history.
The result should actually be better performance than version 0.8.8 while still providing the proper parsing.
Use this link to download a preview of 0.8.10 and and let me know if it works for you:
https://drive.proton.me/urls/SF4MDBBNCG#wZKwjAimMMQv

**Comment by sha-265** [comment]
It's not to old phone (Pixel 4a), and I didn't have performance issues before.

Anyway, I tested this build and it less choppy, but still the transaction history list is still empty:
![image](https://github.com/user-attachments/assets/ed278532-7a16-47d9-9cef-737a38f3e6f1)

In v0.8.8, the history list was loaded before I even go to its screen.

**Comment by michaelWuensch** [comment]
Ok, thanks for testing. If that didn't fix it can only be the change that combined the mpp transactions.
I'll provide another version for you to test in the coming days.

**Comment by michaelWuensch** [comment]
Ok, I was able to reproduce it and found the issue.
The new endpoint I used (that combines mpp payments) does not support pagination for versions older than 24.11.
This means the issue affects all core lightning users with a version older than 24.11 and more than 500 outgoing lightning payments.
I fixed it and still kept the performance improvement I did earlier.
Please test the new version, which you can download below and report back the following:
- Is transaction history loading?
- Are mpp payments combined?
- Are all descriptions for outgoing lightning payments displayed now?

Here is the apk:
https://drive.proton.me/urls/CA9RCY10VG#gNeBXYdcelyJ

**Comment by sha-265** [comment]
Ok, according to my test:
- Yes, the transaction history is loading.
- Yes, MPP payments are combined.
- No, not all descriptions outgoing lightning payments displayed. The same payments' memos as before are missing.

Thank you.

**Comment by michaelWuensch** [comment]
Thanks for testing!
Well, I am out of ideas why some memos are still missing. What really boggles my mind is how Zeus is able to show it. They do nothing fancy in their code.
At least I improved the app while trying to fix this (combined mpp in transaction history, description and payer note added for bolt12 payments & performance improvements)
I am giving up on this until I'm either able to reproduce it myself, or until you have found a pattern or you know how to reproduce it.
Alternatively, as you have access to the payments without memos, you could try to debug and fix it yourself, BitBanana is open source :smile:

**Comment by sha-265** [comment]
Thank you for you efforts. 

> Alternatively, as you have access to the payments without memos, you could try to debug and fix it yourself, BitBanana is open source 😄

I only need to learn Java, and I'm ready to go 🤣

**Comment by sha-265** [comment]
I have an update. After upgrading my CLN node to version 24.11.1, I see incoming transactions, but no outgoing transaction in the history view.

Any idea why?

**Comment by sha-265** [comment]
No. Not on-chain txs at all, I'm talking about regular LN txs.

**Comment by michaelWuensch** [comment]
Oh ok. Actually I wanted to release a new version today. In this version I also reworked the whole pagination in a hopefuly proper way, while it was a quick fix in the last test version I sent you...
So can you please try if the new release would fix it?
https://drive.proton.me/urls/XWFPD322XW#zgGzxzGgzr29
If not, I'll investigate this further before doing the release.

**Comment by sha-265** [comment]
> So can you please try if the new release would fix it?

I tried. Unfortunately the problem still exists.

**Comment by sha-265** [comment]
OK, thanks. I will try the new version with the logs, and I will check if I see something.

**Comment by michaelWuensch** [comment]
Ok, here is the version with an In-App Log. Not the best UX, but I hadn't had that much time ;-)

https://drive.proton.me/urls/J5RYJH6FRW#Mg4uZT1ti4h4

To view the log do the following:
- Go to "Settings -> Features" and enable the Log feature
- Restart the app, so the transaction history is loaded again while Logging is active
- In the main menu find the Log section on the bottom.

The logs are colored by verbosity.
Verbose - gray
Debug - white
Info - blue
Warning - yellow
Error - red

So what you want to look for is most likely yellow or red. If you see some suspicious logs, please let me know!

You can copy a single log line by doing a tap on it or copy all using the "copy all" button if you want to send me the log.

**Comment by sha-265** [comment]
I can't see any thing more then "debugging" in logs:

![Image](https://github.com/user-attachments/assets/7a971a25-633c-40db-9f3a-3d090afd07bd)

**Comment by michaelWuensch** [comment]
That is indeed the part where I expected an error. But it looks exactly the same like on a working instance.
I'll have to add detialed log messages to every single step involved in loading the payments then and export another preview for you.

**Comment by sha-265** [comment]
OK, so when `Watchtower Management` disabled, the history is not working at all, and the log is spammed with this messages:

```
EBUG | CoreLightningApi | Payment page response. Contained elements: 523
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 523
DEBUG | CoreLightningApi | Fetching payments. Pagination turned off.
DEBUG | CoreLightningApi | Payment page response. Contained elements: 523
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 523
DEBUG | CoreLightningApi | Fetching payments. Pagination turned off.
DEBUG | CoreLightningApi | Payment page response. Contained elements: 523
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 523
DEBUG | CoreLightningApi | Fetching payments. Pagination turned off.
DEBUG | CoreLightningApi | Payment page response. Contained elements: 523
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 523
DEBUG | CoreLightningApi | Fetching payments. Pagination turned off.
DEBUG | CoreLightningApi | Payment page response. Contained elements: 523
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 523
DEBUG | CoreLightningApi | Fetching payments. Pagination turned off.
DEBUG | CoreLightningApi | Payment page response. Contained elements: 523
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 523
DEBUG | CoreLightningApi | Fetching payments. Pagination turned off.
DEBUG | CoreLightningApi | Payment page response. Contained elements: 523
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 523
DEBUG | CoreLightningApi | Fetching payments. Pagination turned off.
```

And when it enabled, the history is working (with the bug), and the log shows:

```
DEBUG | CoreLightningApi | Payment page response. Contained elements: 412
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 412
DEBUG | Wallet_TransactionHistory | Payments fetched! Number of LN payments returned:412
DEBUG | Wallet_TransactionHistory | History update finished.
DEBUG | TransactionHistoryFragment | Adding 409 payments to list
DEBUG | TransactionHistoryFragment | Normal payments are displayed
DEBUG | TransactionHistoryFragment | History updated!
DEBUG | CoreLightningApi | Fetching invoices page, offset:  1
DEBUG | CoreLightningApi | Fetching payments page, offset:  1, PageSize: 500
DEBUG | Wallet_Balance | Balances Fetched!
VERBOSE | MainBalanceView | Total balance display updated
DEBUG | CoreLightningApi | Payment page response. Contained elements: 412
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 412
DEBUG | Wallet_TransactionHistory | Payments fetched! Number of LN payments returned:412
DEBUG | Wallet_TransactionHistory | History update finished.
DEBUG | TransactionHistoryFragment | Adding 409 payments to list
DEBUG | TransactionHistoryFragment | Normal payments are displayed
DEBUG | TransactionHistoryFragment | History updated!
```

**Comment by michaelWuensch** [comment]
Okay, thanks for testing.
Sorry my bad the "pagination off" resulted in an endless loop. Didn't really test it as I was in a hurry.

The results from the normal pagination are interesting though. The page size returned is 412 altough you have 523 transactions. Therefore it should be 500.

This brings me to my bext question:
Are really no payments shown or are only the most recent 111 transactions missing? The log says it added the transactions to the list.
So can you scroll all the way down and see if old payments are shown?

**Comment by sha-265** [comment]
Yes, old payments are shown, but I don't know if it exactly 111 txs missing :)

**Comment by michaelWuensch** [comment]
Ok, perfect, then I am pretty sure I know whats going on. Thanks so much for testing!

I'll try to get you another test version this evening that hopefully fixes the issue.

If you are interested in what the issue is, here is my explanation:

I ask CLN for payments with a page size of 500 AND only want to get completed payments. 
I expected it to return 500 if there are more than 500 but it seems to check 500 payments and then remove the not completed ones. In your example 412 are left and returned.
My code continues to load pages until the returned pagesize is smaller than the requested one. This is already the case in the first response though.
On my regtest node I probably had only completed payments, therefore it worked.

**Comment by sha-265** [comment]
Interesting. Thanks for the hard work to fix it!

**Comment by michaelWuensch** [comment]
Ok, here it is.
https://drive.proton.me/urls/X8TW2R4QJ8#cpisEPCcIyT9
Please let me know if it works now.
Even if it works, it would be cool if you could post the log so I can verify everything behaves like expected :)

**Comment by sha-265** [comment]
Now even more payments are missing 😅 The last payment I see is older now

**Comment by michaelWuensch** [comment]
Oh come on! Sorry.
Can you send the log?

**Comment by sha-265** [comment]
I don't why, but the numbers are completely different:
```
DEBUG | CoreLightningApi | Fetching invoices page, offset:  1
DEBUG | CoreLightningApi | Fetching payments page, offset:  1, PageSize: 500
DEBUG | Wallet_Balance | Balances Fetched!
VERBOSE | MainBalanceView | Total balance display updated
DEBUG | Wallet_Bolt12Offers | bolt12 offers fetched!
DEBUG | Wallet_Channels | Channels fetched!
DEBUG | AliasManager | Loaded Alias cache.
DEBUG | Wallet_Channels | Fetching node info for 0 nodes.
INFO | Wallet | New wallet load state: WALLET_LOADED
VERBOSE | HomeActivity | Scheduled node info check initiated.
VERBOSE | ClipBoardUtil | Clipboard with same content was checked before
DEBUG | CoreLightningApi | Payment page response. Contained elements: 316
DEBUG | CoreLightningApi | Final returned payment page. Contained elements: 289
DEBUG | Wallet_TransactionHistory | Payments fetched! Number of LN payments returned:289
DEBUG | Wallet_TransactionHistory | History update finished.
DEBUG | TransactionHistoryFragment | Adding 289 payments to list
DEBUG | TransactionHistoryFragment | Normal payments are displayed
DEBUG | TransactionHistoryFragment | History updated!
```

**Comment by michaelWuensch** [comment]
Ok, I can't make sense of this.
Now I said:
Give me 500 transactions no matter if they are completed or not... And it returns less than before.
Sigh. I guess I'll increase the Page Size to a crazy big number like 5000 for now which should "fix" it for you, do the release and then I have to find a way to replicate it on my regtest node. Finding a bug with our current procedure is verry time consuming for both of us.
But its still great we now have a Log feature in BitBanana! 😀

**Comment by michaelWuensch** [comment]
I just released the official version 0.9.0.
I was also able to reproduce the issue and added a fix that at least fixed it for me.
Please let me know if in the new official release all Payments are back.
And if they are, could you also check again if the original issue still exists (memos not displayed).

**Comment by sha-265** [comment]
Yes, now I see all the latest payments. Thanks. What was the problem?

But the original issue about the memos still exists 😅

**Comment by michaelWuensch** [comment]
The issue was that a page size of 500 means 500 payment parts. MPP payments get combined in the response though, so less then 500 elements get returned which stopped requesting the next page.
In short the problem only arrised for nodes with successfull mpp payments, which my test node did not have until yesterday.

Anyway, glad to hear it now works.

I'll still keep the issue open, as the original issue still persists.
Please let me know if you have any additional info for the original issue or if you find any procedure on how to reproduce it.
I am still clueless and can't reproduce it.

**Comment by sha-265** [comment]
I don't have other ideas. I tried to decode both working and not working memos BOLT11, and they look the same. But this bug annoying :(

**Comment by michaelWuensch** [comment]
I unfortunately still did not encounter such a case, which makes it very hard to find a solution. :(

**Comment by michaelWuensch** [comment]
As you have the node that has the issues, you could debug it yourself by building and modifying the code. If you are interested, I can write you a short summary on how to do it and link you all relevant lines of code, so you don't have to search forever.

**Comment by sha-265** [comment]
Even though I never worked with java, I can try maybe with good instructions :)

**Comment by michaelWuensch** [comment]
Cool, thanks.
I am on the road right now, but first steps would be to build directly to your phone.
Here are the instructions to do so:
https://github.com/michaelWuensch/BitBanana/blob/master/docs/INSTALL.md#build-directly-to-your-phone

If you get stuck somewhere or need more details, let me know.

If it works, you will have an additional BitBanana app on your smartphone called "BitBanana Debug" where the logo background is black instead of blue.
This is so you can keep your normal BitBanana setup while experimenting / debugging / coding new features.

You can create a backup in your BitBanana app and then restore it inside of BitBanana debug. This way you don't need to go through the connection process again.

**Comment by michaelWuensch** [comment]
@sha-265 
When you managed to get it to build you can start debugging.
Before finding the issue why the data is not displayed, we first have the find the data at all.

This is the code of interest:
https://github.com/michaelWuensch/BitBanana/blob/365cf563a9f10ae600c4186f61c0823d5a9e6088/app/src/main/java/app/michaelwuensch/bitbanana/backends/coreLightning/CoreLightningApi.java#L734-L763

To easier debug this I would suggest you add a little bit of code right between the following 2 lines:
https://github.com/michaelWuensch/BitBanana/blob/365cf563a9f10ae600c4186f61c0823d5a9e6088/app/src/main/java/app/michaelwuensch/bitbanana/backends/coreLightning/CoreLightningApi.java#L739-L740

In BitBanana go to the payment where the description is missing, then copy the preimage.
In the code then add something like this:

``` Java
if (ApiUtil.StringFromHexByteString(payment.getPreimage().equals("INSERT_YOUR_COPIED_PREIMAGE_HERE")){
    BBLog.d("DESCRIPTION_BUG", "This is the description for the payment: " + payment.getDescription());
    BBLog.d("DESCRIPTION_BUG", "This is the label for the payment: " + payment.getLabel());
    BBLog.d("DESCRIPTION_BUG", "This is the description extracted from bolt 11: " + InvoiceUtil.getBolt11DescriptionFast(payment.getBolt11()));
}
```
The if clause makes sure that stuff gets only executed for the transaction you want to debug. This allows you the add lots of debug messages without spamming your whole log.
While the phone is still connected to your PC, you can use Android Studios' LogCat to display the logs. Filter the logs to "BBLog" if you want to display all BitBanana Logs. Filter it to "DESCRIPTION_BUG" if you only want to diplay your logs done in that if clause.

In the example code I added 3 logs where the information might be found. If any of these 3 logs returns the missing description we can look where in the code it gets lost before displaying it. If none of the 3 logs has the info, the I don't know where the hell Zeus has the info from. (Unless it would be a bug in the grpc implementation and only the Rest implementation returns the info.)

Please let me know what you find out or if you need further help.
You can also always ask chatGPT questions about the code. It is really useful if you don't understand a specific programming language and can explain it very well to you.

**Comment by michaelWuensch** [comment]
@sha-265 Can you give me the preimage hex string of a payment where the memo/description is not displayed?
You can obtain it by going to the transaction details and then copy the preimage.
I'll build a version that will display logs for that transaction and provide it to you. Maybe that helps us to finally fix the issue.

**Comment by sha-265** [comment]
I will send you on Discord.

BTW, now I see the memos are missing for these txs also in Zeus 😵

**Comment by sha-265** [comment]
OK, I did it, and now I see the rebalance is a keysend tx 🥴 I can't find the old ones.

So it's doesn't have BOLT11 at all, and the memo come from there in other txs, as I understand. 

I guess memos for keysend should be some where else? Because it still worked with Zeus previously somehow.

**Comment by michaelWuensch** [comment]
Ah, okay. Displaying keysend messages is not supported, as the Core Lightning API does not expose the needed information to read it. I have created an issue for that more than a year ago: https://github.com/ElementsProject/lightning/issues/7169

My best guess is that the 3rd party c-lightning Rest plugin that Zeus used until 0.9.0 maybe exposed that and now that they switched to the official CLNRest plugin they probably don't have access to that data as well.
